java-checks/src/main/java/org/sonar/java/checks/LazyArgEvaluationCheck.java

```diff
@@ -0,0 +1,287 @@
+      WARN,
+      ERROR;
+
+      final List<MethodMatcher> log = varArgVariants(SLF4J_LOGGER, toString().toLowerCase(Locale.ROOT));
+
+      final MethodMatcher test = levelTestProto(SLF4J_LOGGER).name(String.format("is%c%sEnabled", toString().charAt(0), toString().toLowerCase(Locale.ROOT).substring(2)));
+
+      @Override
+      public List<MethodMatcher> log() {
+        return log;
+      }
+
+      @Override
+      public MethodMatcher test() {
+        return test;
+      }
+    }
+
+    enum JUL_LEVELS implements LogLevels {
+      SEVERE,
+      WARNING,
+      INFO,
+      CONFIG,
+      FINE,
+      FINER,
+      FINEST;
+
+      final List<MethodMatcher> log = ImmutableList.of(loggerProto(JUL_LOGGER).name(toString().toLowerCase(Locale.ROOT)));
+
+      final MethodMatcher test = levelTestProto(JUL_LOGGER).name(String.format("is%c%sEnabled", toString().charAt(0), toString().toLowerCase(Locale.ROOT).substring(2)));
+
+      @Override
+      public List<MethodMatcher> log() {
+        return log;
+      }
+
+      @Override
+      public MethodMatcher test() {
+        return test;
+      }
+    }
+  }
+
+  private static final MethodMatcher PRECONDITIONS = MethodMatcher.create()
+    .typeDefinition("com.google.common.base.Preconditions")
+    .name("checkState")
+    .withAnyParameters();
+
+  private static final MethodMatcher JUL_LOG = MethodMatcher.create()
+    .typeDefinition(JUL_LOGGER)
+    .name("log")
+    .addParameter(TypeCriteria.anyType())
+    .addParameter(String.class.getCanonicalName());
+
+  private static final MethodMatcherCollection LAZY_ARG_METHODS = MethodMatcherCollection.create(PRECONDITIONS, JUL_LOG);
+  static {
+    LogLevels.logLevels().forEach(l -> LAZY_ARG_METHODS.addAll(l.log()));
+  }
+
+  private static final MethodMatcherCollection LOG_LEVEL_TESTS = MethodMatcherCollection.create();
+  static {
+    LogLevels.logLevels().forEach(l -> LOG_LEVEL_TESTS.add(l.test()));
+  }
+
+  private JavaFileScannerContext context;
+  private Deque<Tree> treeStack = new ArrayDeque<>();
+
+  @Override
+  public void scanFile(JavaFileScannerContext context) {
+    this.context = context;
+    if (context.getSemanticModel() == null) {
+      return;
+    }
+    scan(context.getTree());
+  }
+
+  @Override
+  public void visitMethodInvocation(MethodInvocationTree tree) {
+    if (LAZY_ARG_METHODS.anyMatch(tree) && !insideCatchStatement() && !insideLevelTest()) {
+      onMethodInvocationFound(tree);
+    }
+  }
+
+  @Override
+  public void visitIfStatement(IfStatementTree ifTree) {
+    Slf4jTestVisitor slf4jTestVisitor = new Slf4jTestVisitor();
+    ifTree.condition().accept(slf4jTestVisitor);
+    if (slf4jTestVisitor.match) {
+      stackAndContinue(ifTree, super::visitIfStatement);
+    }
+  }
+
+  @Override
+  public void visitCatch(CatchTree tree) {
+    stackAndContinue(tree, super::visitCatch);
+  }
+
+  @Override
+  public void visitMethod(MethodTree tree) {
+    stackAndContinue(tree, super::visitMethod);
```

Comment:
May worth a comment why you need to stack a method, as you never check for it in the stack, even if I guess it's done in order to avoid FPs in anonymous classes.